### PR TITLE
add option to specify network interface

### PR DESCRIPTION
**Pull request by parrasajad**
## Proposed changes
Example:
```console
echo https://example.com | nuclei -i en0
```

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: v2/cmd/nuclei/main.go [hunk 1]
```diff
@@ -160,6 +160,7 @@ on extensive configurability, massive extensibility and ease of use.`)
 		flagSet.BoolVarP(&options.ShowMatchLine, "show-match-line", "sml", false, "show match lines for file templates, works with extractors only"),
 		flagSet.BoolVar(&options.ZTLS, "ztls", false, "use ztls library with autofallback to standard one for tls13"),
 		flagSet.StringVar(&options.SNI, "sni", "", "tls sni hostname to use (default: input domain name)"),
+		flagSet.StringVarP(&options.Interface, "interface", "i", "", "network interface to use for network scan"),
```
Comment (Mzack9999) [inline v2/cmd/nuclei/main.go:163]: I think the user should specify either:
- Source Ip (we need to ensure that at least one interface exists with such source IP)
- Interface and interface's Source IP (we need to ensure that the specified interface has the desired source IP)

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: v2/pkg/protocols/common/protocolstate/state.go [hunk 2]
```diff
@@ -16,6 +19,48 @@ func Init(options *types.Options) error {
...
+			return err
+		}
+		if isAssociated {
+			opts.Dialer = &net.Dialer{
+				LocalAddr: &net.TCPAddr{
+					IP: net.ParseIP(options.SourceIP),
+				},
+			}
+		} else {
+			return fmt.Errorf("sourceIP (%s) is not associated with the interface (%s)", options.SourceIP, options.Interface)
```
Comment (Mzack9999) [inline v2/pkg/protocols/common/protocolstate/state.go:36]: `sourceIP` => `source ip`

File: v2/pkg/protocols/common/protocolstate/state.go [hunk 3]
```diff
@@ -16,6 +19,48 @@ func Init(options *types.Options) error {
...
+			return err
+		}
+		if isAssociated {
+			opts.Dialer = &net.Dialer{
+				LocalAddr: &net.TCPAddr{
+					IP: net.ParseIP(options.SourceIP),
+				},
+			}
+		} else {
+			return fmt.Errorf("sourceIP (%s) is not associated with any network interface", options.SourceIP)
```
Comment (Mzack9999) [inline v2/pkg/protocols/common/protocolstate/state.go:50]: `sourceIP` => `source ip`

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
Requesting minor changes (string rename)

**Review by Mzack9999 — APPROVED** [review]

**Review by ehsandeep — CHANGES_REQUESTED** [review]
- [x] Crash upon random interface input

@parrasajad 

```console
./nuclei -i ll

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   2.7.5-dev

		projectdiscovery.io

[WRN] Use with caution. You are responsible for your actions.
[WRN] Developers assume no liability and are not responsible for any misuse or damage.
[INF] Using Nuclei Engine 2.7.5-dev (development)
[INF] Using Nuclei Templates 9.1.5 (latest)
[INF] Templates added in last update: 58
[INF] Templates loaded for scan: 3835
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x105227448]

goroutine 659 [running]:
github.com/projectdiscovery/fastdialer/fastdialer.(*Dialer).DialTLS(0x0, {0x1061c4c50, 0x14004436d20}, {0x1057a9447, 0x3}, {0x140057c6bd0, 0x12})
	/Users/geekboy/go/pkg/mod/github.com/projectdiscovery/fastdialer@v0.0.16-0.20220609092737-7e3c7dcecf16/fastdialer/dialer.go:106 +0x38
net/http.(*Transport).customDialTLS(0xfe7?, {0x1061c4c50?, 0x14004436d20?}, {0x1057a9447?, 0x140066afa38?}, {0x140057c6bd0?, 0x140066afa08?})
	/opt/homebrew/Cellar/go/1.18.2/libexec/src/net/http/transport.go:1317 +0x54
net/http.(*Transport).dialConn(0x140002b5400, {0x1061c4c50, 0x14004436d20}, {{}, 0x0, {0x14003812090, 0x5}, {0x140057c6bd0, 0x12}, 0x0})
	/opt/homebrew/Cellar/go/1.18.2/libexec/src/net/http/transport.go:1583 +0x2bc
net/http.(*Transport).dialConnFor(0x14005f0a240?, 0x14004c7e4d0)
	/opt/homebrew/Cellar/go/1.18.2/libexec/src/net/http/transport.go:1449 +0x80
created by net/http.(*Transport).queueForDial
	/opt/homebrew/Cellar/go/1.18.2/libexec/src/net/http/transport.go:1418 +0x39c
```

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**